src/sentry/api/authentication.py

```diff
@@ -69,6 +70,40 @@ def authenticate_credentials(self, userid, password):
         return (AnonymousUser(), key)
 
 
+class ClientIdSecretAuthentication(QuietBasicAuthentication):
+    """
+    Authenticates a Sentry Application using its Client ID and Secret
+
+    This will be the method by which we identify which Sentry Application is
+    making the request, for any requests not scoped to an installation.
+
+    For example, the request to exchange a Grant Code for an Api Token.
+    """
+
+    def authenticate(self, request):
+        if not request.json_body:
+            raise AuthenticationFailed('Invalid request')
+
+        client_id = request.json_body.get('client_id')
+        client_secret = request.json_body.get('client_secret')
+
+        if not client_id or not client_secret:
+            raise AuthenticationFailed('Missing Client ID or Secret')
+
+        try:
+            application = ApiApplication.objects.get(client_id=client_id)
+        except ApiApplication.DoesNotExist:
+            raise AuthenticationFailed('Invalid Client ID')
+
+        if not constant_time_compare(application.client_secret, client_secret):
+            raise AuthenticationFailed('Invalid Client ID / Secret pair')
```

Comment:
Should we expose the fact that the clientid matches but the secret is wrong? Generally in password based scenarios you don't want to divulge that information.

Reply:
I worded it like this to keep it a little vague (ie. "either one could wrong, but we won't say which one")

Reply:
But an invalid client id will give `Invalid Client ID`. This error message is different and lets a :imp: know they have moved onto a new condition.

Reply:
oh, heh. Whoops. Good call.

Reply:
I just updated this to return the same generic "Invalid pair" error, regardless of the cause.